src/annotator/frame-observer.js

```diff
@@ -65,22 +53,71 @@ export default class FrameObserver {
    */
   _removeFrame(frame) {
     this._onFrameRemoved(frame);
-
-    // Remove the frame from our list
-    this._handledFrames = this._handledFrames.filter(x => x !== frame);
+    this._handledFrames.delete(frame);
   }
 
   _discoverFrames() {
-    let frames = FrameUtil.findFrames(this._element);
+    let frames = findFrames(this._element);
 
     for (let frame of frames) {
-      if (!this._handledFrames.includes(frame)) {
+      if (!this._handledFrames.has(frame)) {
         this._addFrame(frame);
       }
     }
 
-    for (let frame of difference(this._handledFrames, frames)) {
-      this._removeFrame(frame);
+    for (let frame of this._handledFrames) {
+      if (!frames.has(frame)) {
+        this._removeFrame(frame);
+      }
     }
   }
 }
+
+/**
+ * Check if we can access this iframe's document
+ *
+ * @param {HTMLIFrameElement} iframe
+ */
+function isAccessible(iframe) {
+  try {
+    return !!iframe.contentDocument;
+  } catch (e) {
+    return false;
+  }
+}
+
+/**
+ * Resolves a Promise when the iframe's DOM is ready (loaded and parsed)
+ *
+ * @param {HTMLIFrameElement} iframe
+ * @return {Promise<void>}
+ */
+export function onDocumentReady(iframe) {
+  return new Promise(resolve => {
+    const iframeDocument = /** @type {Document} */ (iframe.contentDocument);
+    if (iframeDocument.readyState === 'loading') {
+      iframeDocument.addEventListener('DOMContentLoaded', () => {
+        resolve();
+      });
+    } else {
+      resolve();
+    }
+  });
+}
+
+/**
+ * Return all `<iframe>` elements under `container` which are annotate-able.
+ *
+ * To enable annotation, an iframe must be opted-in by adding the
+ * `enable-annotation` attribute.
+ *
+ * Eventually we may want annotation to be enabled by default for iframes that
+ * pass certain tests. However we need to resolve a number of issues before we
+ * can do that. See https://github.com/hypothesis/client/issues/530
+ *
+ * @param {Element} container
+ * @return {Set<HTMLIFrameElement>}
+ */
+export function findFrames(container) {
```

Comment:
`findFrames` is only used internally within `FrameObserver` and the implementation is pretty trivial. It is also inherently coupled to `FrameObserver` because the CSS selector here and the `MutationObserver` configuration need to be in-sync. I think it would make sense to inline the logic into `discoverFrames` and move the documentation here to the `FrameObserver` class.

Reply:
👍🏼